feat: Make MonitorUpdatingPersister change persist type based on size

Prabhat1308 opened the pull request:
fixes #3770 

- Skips full persistence when ChannelMonitor is smaller than a pre-determined size.
- Adds a field `minimum_monitor_size_for_updates` to specify the minimum size for full persistence to be activated
- Adds `new_with_default_threshold` function to setup `MonitorUpdatingPersister` with a default `minimum_monitor_size_for_updates` value

tnull reviewed (commented):
Mostly looks good, but can we add some test coverage for the new behavior?

Additionally, benchmarks would indeed very helpful to evaluate what a reasonable threshold value would be.

Prabhat1308 commented on lightning/src/util/persist.rs:
```diff
@@ -491,21 +499,45 @@ where
+	/// in full rather than using update-based persistence.
+	///
+	/// For other parameters, see [`MonitorUpdatingPersister::new`].
+	pub fn new_with_default_threshold(
```
done.

tnull reviewed (commented):
Please rewrite your commit history so that it's clear what are feature and fixup commits. Each commit message should have a clear headline followed by some paragraph(s) describing the change, where it happened, and why it was necessary. For guidance, please take a look at https://cbea.ms/git-commit/

Prabhat1308 committed:
Add size-based optimization to MonitorUpdatingPersister

Introduces an optimization to the MonitorUpdatingPersister to
avoid writing differential updates for small channel monitors.

When a channel monitor is smaller than a configurable threshold
, the persister will now write the full monitor instead of an update.
This avoids the I/O overhead of creating and managing many small update
files for monitors that don't benefit significantly from differential updates.

Prabhat1308 committed:
Add tests for persister size-based optimization

Adds unit test for introduced size based optimisation. Also updates
the old unit test to use a threshold value and use constructors to
increase test coverage

Prabhat1308 committed:
change threshold value

changes the threshold value to 8192 bytes as suggested in the PR comments

domZippilli commented:
[Left some thoughts on the issue](https://github.com/lightningdevkit/rust-lightning/issues/3770#issuecomment-2984271612), since my thoughts aren't about this implementation but more the feature itself. But posting here since this is where the recent action is. 🙇

tnull reviewed (commented):
Current changes look good to me, but I'll defer ACKing to after the discussion  on whether we want this change afterall has concluded.

domZippilli commented on lightning/src/util/persist.rs:
```diff
@@ -491,14 +499,45 @@ where
+			kv_store,
+			logger,
+			maximum_pending_updates,
+			8192,
```
Echoing @tnull's request to make this an appropriately named constant from a previous diff. https://github.com/lightningdevkit/rust-lightning/pull/3834#discussion_r2171165554

Prabhat1308 replied:
added the named const.

domZippilli commented on lightning/src/util/persist.rs:
```diff
@@ -491,14 +499,45 @@ where
+	/// monitors. Set to 0 to always use update-based persistence regardless of size.
+	///
+	/// For other parameters, see [`MonitorUpdatingPersister::new`].
+	pub fn new_with_monitor_size_threshold(
```
I wonder if, instead of two constructors, it would be nicer/better to either add this to the default constructor, or use a config struct, or write a builder. 

It's not a problem here, but I wonder if we'll think of another tunable for MUP and then there will be a third constructor, and so on.

Prabhat1308 committed:
set a default const and maintain single constructor

changes the code to have a proper constant value than a unnamed hardcoded constant . Removes the extra constructor added and adjust the min_monitor_size_for_updates_bytes variable in the constructor
